release 9.0 (#222)
reason-association/rescript-lang.org: _blogposts/2020-02-09-release-9-0.mdx

---
author: hongbo

> date: "2020-02-09" 

Reviewer: ```suggestion
date: "2021-02-09" 
```

> previewImg:

Reviewer: I will create a release preview image template for the future to make it easier to spot in the blog overview later on

Author: sounds great, I can wait until tomorrow

category: compiler

> title: ReScript 9.0

Reviewer: ```suggestion
title: ReScript 9.0
badge: release
```

This will show a "Release" ribbon on your blog image

description: |
---

## Introduction

> ReScript is a soundly typed language with an optimizing compiler focused on the JS platform. 

Reviewer: It's hard to review with the line breaks in each paragraph, can you remove the line breaks, so that every paragraph is soft-wrapped on one line?

Author: I am not sure if I follow you, you mean no new line for each paragraph?

It's focused on type safety, performance and JS interop. It used to be called BuckleScript.

[ReScript@9.0.0](https://www.npmjs.com/package/bs-platform/v/9.0.0) is now available, you can try it via

```
npm i bs-platform@9.0.0
```

The changes are listed [here](https://github.com/rescript-lang/rescript-compiler/blob/master/Changes.md#90). 

We will go through some highlighted user visible changes.

### Customized light weight stdlib

This is a long-awaited [feature request](https://github.com/rescript-lang/rescript-compiler/pull/2171). 

> Since this release, users can add such lines in bsconfig.json to make our compiler pakcage a dev dependency.

Reviewer 2: ```suggestion
With this release, users can add the following line in bsconfig.json to make our compiler package a dev dependency.
```

```json
"external-stdlib" : "@rescript/std"
```

Our current compiler package is quite large since it ships both prebuilt compilers for 3 major platforms 
and the standard library. With this configuration, the compiler will generate js files on top of `@rescript/std` so that
the compiler package is not needed after the JS files are generated. This is helpful if you want to export ReScript based 
libraries to JS users while not asking JS users to install a large dependency.

Caveat: make sure the version of `@rescript/std` is the same as `bs-platform`, use this configuration only when the package size 
is indeed a problem for you.

### Zero-cost bundle size when adding ReScript 

> We keep generating code that is friendly for tree-shaking during each releases. 
> We believe we reach a milestone that ReScript adds close to zero cost to your bundler 
> size unlike many other programming languages compiled into JS; the bundled code is almost ReScript runtime free and 
> the generated library code is fits the tree-shaking principle very well.

Reviewer 2: ```suggestion
With each release, we keep a close eye on generating code that is optimized for tree-shaking. 
We believe we reached a milestone that ReScript adds close to zero cost to your bundle-size.
Unlike many other programming languages compiled into JS, the bundled code is almost ReScript runtime free and 
the generated library code fits the tree-shaking principle very well.
```

To demonstrate what we achieved, we made a [repo](https://github.com/bobzhang/zero-cost-rescript) so that 
you can try it and see how good the bundled code is.

### Improved code generation for pattern match

We continue improving the quality of generated code in this release.

> Take this [issue](https://github.com/rescript-lang/rescript-compiler/issues/4924) for example, for pattern match as below:

Reviewer 2: ```suggestion
Take this [issue](https://github.com/rescript-lang/rescript-compiler/issues/4924) for example:
```

```res
let test = x =>
  switch x {
  | NoArg => true
  | _ => false
  }
```

> It used to generate code as below:

Reviewer 2: ```suggestion
It used to generate the following code:
```

```js
function test(x) {
  if (typeof x === "number") {
    return x === 0;
  } else {
    return false;
  }
}

```

> Now it is simplified as 

Reviewer 2: ```suggestion
which now gets optimized to: 
```

```js
function test(x){
  return x === 0
}
```

> This is possible is that our optimizer will try to analyze several predicates and get rid of redundant predicates, 
> more diffs could be found [here](https://github.com/rescript-lang/rescript-compiler/pull/4927/files?file-filters%5B%5D=.js).

Reviewer 2: ```suggestion
This is possible because our optimizer will try to analyze several predicates and get rid of redundant ones. 
More diffs can be found [here](https://github.com/rescript-lang/rescript-compiler/pull/4927/files?file-filters%5B%5D=.js).
```

> Another important improvement is that we fixed the pattern match offset issue, the consequence is that
>  magic number will not be generated
> for complex pattern matches any more, below is a represenative diff thanks to this clean up:

Reviewer 2: ```suggestion
Another important improvement is that we fixed the pattern match offset issue, which leads to the consequence, that magic numbers will not be generated for complex pattern matches anymore. 
Below is a representative diff resulting from this cleanup:
```

```diff
function is_space(param){
- var switcher = param - 9 | 0;
- if (switcher > 4 || switcher < 0) {
-    return switcher == 23 ; 
+ if (param > 13 || param < 9) {
+    return param === 32;
  } else {
-    return switcher !== 2;     
+    return param != 11;
  }    
}
```

### Syntax changes

We introduced two minor tweaks for the concrete syntax.

> - guard in pattern matching is recommended to use `if` instead of `when`

Reviewer 2: ```suggestion
- It is now recommended to use `if` instead of `when` in pattern match guards
```

```res
switch expr {
  | pat if predicate => // was: pat when predicate  
}
```

`when` is still supported but it will be deprecated in the future.

- explicit nested object literal syntax

> `{"keya" : {keyb : 3 } }` used to be interpreted as `{"keya" : {"keyb" : 3 } } `, such implicity makes 
> us not able to embed regular records inside JS object literals. Since this release, 
> user has to write JS object literals explicitly and `{"keya" : {keyb : 3}}` will be interpreted as a regular 
> record inside a js object literal. More discussions can be found [here](https://forum.rescript-lang.org/t/fixing-the-semantics-of-nested-objects-breaking-changes/976).

Reviewer 2: ```suggestion
`{ "keya": { keyb: 3 } }` used to be interpreted as `{ "keya": { "keyb": 3 } } `, such implicity makes 
us not able to embed regular records inside JS object literals. With this release, 
JS object literals have to be written explicitly and `{ "keya": { keyb: 3 } }` will be interpreted as a regular 
record inside a JS object literal. More discussions can be found [here](https://forum.rescript-lang.org/t/fixing-the-semantics-of-nested-objects-breaking-changes/976).
```

> Note there are some pretty interesting internal changes in this release -- for example, using WASM to replace Camlp4 and a 
> generalized visitor pattern without using objects -- that we will discuss in a separate post

Reviewer 2: ```suggestion
There are also some pretty interesting internal changes in this release.
For example, using WASM to replace Camlp4 and a generalized visitor pattern without using objects, which we will discuss in a separate post.
```
